### ISSUE TITLE
systemd in new namespace masks /run bound into a Singularity container

### ISSUE THREAD

**Issue by Heng-Zhou** [sylabs/singularity#2656]
**Version of Singularity**
```
$ singularity-ce version 4.0.2-jammy
```

**Describe the bug**
Mount point is visible in new namespace if created in normal Ubuntu, but invisible if created in Singularity container.

**To Reproduce**
1) Create a new namespace with command `/usr/bin/daemonize /bin/unshare -Urfp --mount-proc bash -c 'exec /lib/systemd/systemd --unit=basic.target'`.
2) Find out the PID for process `/lib/systemd/systemd --unit=basic.target` in results returned from command `lsns |grep systemd`. Save it in `SYSTEMD_PID`.
3) Enter the newly created child namespace with command `/usr/bin/nsenter --preserve-credentials --user --mount --pid -t $SYSTEMD_PID`.

Run the above three same commands on normal Ubuntu and inside a Singularity container containing Ubuntu, respectively.

The definition file `test.def` of the Singularity container is:
```
BootStrap: library
From: ubuntu

%post
    set -xe
    apt-get -y update && apt -y upgrade
    apt install -yqq daemonize dbus-user-session fontconfig

%environment
    export LC_ALL=C
```
And the container is shelled in by command `singularity shell --fakeroot --userns --writable-tmpfs --bind /run/user/$(id -u)/ test.sif`.

**Expected behavior**
On normal Ubuntu, the directory indicated by `DBUS_SESSION_BUS_ADDRESS` -- `/run/user/<my_host_UID>` can be seen in the new namespace.
On Singularity container, however, `/run/user/<my_host_UID>` is invisible in the new namespace. 

`/run/user/<my_host_UID>` is a mount point in both scenarios, which can be checked by `mount` command.

Since the commands 1) -3) are the same, people should expect the results are the same. That is, we should see the `DBUS_SESSION_BUS_ADDRESS` `/run/user/<my_host_UID>` all the same in the second case. This expectation is more reasonable if considering the fact that the directory is mounted as a mount point in both cases. But this directory is invisible in the new namespace created in Singularity container. The only difference between the two scenarios is that the latter is in Singularity container. So, I have to think this is a bug of Singularity.

**OS / Linux Distribution**
Which Linux distribution are you using?
```
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

**Installation Method**
1) Download the .deb file from github repository. 
2) Install the downloaded .deb file using `dpkg -i` command.

**Comment by dtrudg** [comment]
> The only difference between the two scenarios is that the latter is in Singularity container. So, I have to think this is a bug of Singularity.

This is not due to Singularity, but is due to the different behaviour of systemd when it is run inside the container.

If you examine inside of your systemd namespace, you will find that the `/run` tmpfs there has been *created and mounted by systemd*. This is masking the bind mount from the host that singularity is providing.

In the systemd namespace note the contents and times - this is a new /run tmpfs that systemd created:

```
root@mini:/run
09:44 AM $ ls -lah
total 0
drwxr-xr-x. 5 root root 100 Feb 13 09:40 ./
drwxr-xr-x. 1 root root  60 Feb 13 09:40 ../
drwxr-xr-x. 2 root root  40 Feb 13 09:40 credentials/
drwxrwxrwt. 2 root root  40 Feb 13 09:40 lock/
drwxr-xr-x. 4 root root  80 Feb 13 09:40 systemd/
```

If I just use the unshare command inside the container to run bash, I can still see the /run from the host. Singularity itself is not causing an issue here as the bind mount is still visible inside my newly create namespace in the container:

```
Singularity> /bin/unshare -Urfp --mount-proc bash
root@mini:/home/dtrudg-sylabs
09:45 AM $ ls -lah /run
total 0
drwxr-xr-x.  1 root root  60 Feb 13 09:40 ./
drwxr-xr-x.  1 root root  60 Feb 13 09:40 ../
drwxrwxrwt.  3 root root  60 Feb 13 09:37 lock/
drwxr-xr-x.  2 root root  40 Feb 13 09:37 log/
drwxr-xr-x.  2 root root  40 May 31  2022 mount/
drwxr-xr-x.  2 root root  40 Feb 13 09:37 sendsigs.omit.d/
lrwxrwxrwx.  1 root root   8 Feb 13 09:37 shm -> /dev/shm/
drwxr-xr-x. 10 root root 220 Feb 13 09:37 systemd/
drwxr-xr-x.  1 root root  60 Feb 13 09:40 user/
```

I cannot tell you precisely why the behaviour of systemd differs, but the environment it sees is very different between a namespace created on the Ubuntu host system and that inside the singularity container. On the Ubuntu host it will see the host machine ID, systemd configuration, unit files etc. In the container it will not see these.

I would not expect it to be possible to share `/run` on the host with a completely separate systemd in the container. The state on the host and the container will be different, and need to be different.

Fully supporting systemd inside a container automatically is complex, requiring the container runtime to setup certain things (see e.g. https://systemd.io/CONTAINER_INTERFACE/). This is out of scope for SingularityCE - we explictly aim to provide *application containers*, while systemd is a component of a full operating system image.

I'm not clear of your overall goal here, but the complexities of this approach aren't something we can continue to offer support for withing the limitations of an open source project.

I would advise, instead, stating the end goals (e.g. "I want to run the GNOME desktop in a container") in a GitHub discussions / Google Groups thread, or in our Slack workspace, and being clear about what you have tried. You may find other SingularityCE users who have a working approach that does not rely on createing a full operating system container running systemd.

**Comment by Heng-Zhou** [comment]
I figured out the cause of `/run/user/<my_host_UID>` not showing. At first glance, it is because systemd created its own /run. When `ls /run/user/<my_host_UID>` is executed, the command traverses from top to bottom: / -> /run/ -> /run/user -> /run/user/<my_host_UID>. It terminates, however, at the 3rd step because `/run/user` does not exist, leading to `/run/user/<my_host_UID>` not shown.

Creating `/run` is systemd's inherent function (see [this](https://unix.stackexchange.com/questions/391555/how-to-mount-run-early-enough) for example), if this directory does not exist. In my question, I only bind mount `/run/user/<my_host_UID>`, so /run does not exist when systemd is running, and systemd hence creates and mounts its own /run, which does not contains user subdirectory.

But if `/run` exists when systemd is running, systemd will not create its own /run and mount it (so that the original /run would not be replaced, perhaps for the purpose of preserving the current system settings). So, if I bind only `/run` in the command that shells into the container instead of `--bind /run/user/$(id -u)/`, /run will exist when systemd is running, so systemd would not create and mount its own /run. Singularity seems to have a nice feature -- it will mount all mount points in the subdirectory. So, with /run being mounted, its sub mount point `/run/user/<my_host_UID>` will automatically be mounted too, which will in turn be brought into the new namespace. As a result, `ls /run/user/<my_host_UID>` command will work as expected.

To sum up, it has nothing to do with "mask". The problem is systemd creating and mounting its own /run when /run does not exist because it is not bound into the Singularity container.

**Comment by dtrudg** [comment]
> I only bind mount /run/user/<my_host_UID>, so /run does not exist when systemd is running, and systemd hence creates and mounts its own /run, which does not contains user subdirectory.

> To sum up, it has nothing to do with "mask". The problem is systemd creating and mounting its own /run when /run does not exist because it is not bound into the Singularity container.

I'm glad that you have a solution, but I would caution against drawing this conclusion as the reason for the issue.

If you are binding in `/run/user/<uid>` into the container then `/run` and `/run/user` _must_ be present in the container, or there is nowhere for singularity to bind the `<uid>` directory onto. In the absence of those directories in the container rootfs, Singularity will create them in an overlay or underlay session.

As mentioned, there is a lot of complexity involved with running systemd in a container. I'll close this issue, and if you meet further problems I would ask that you use the Q&A discussions as I mentioned previously.

**Comment by Heng-Zhou** [comment]
> If you are binding in `/run/user/<uid>` into the container then `/run` and `/run/user` _must_ be present in the container, or there is nowhere for singularity to bind the `<uid>` directory onto. In the absence of those directories in the container rootfs, Singularity will create them in an overlay or underlay session.

Yes, `/run` and `/run/user` are present as directory in the container. But the key here is that `/run` is not present as a mount point in the container, thus systemd creates its own and mounts it.

PS, I have finished the GNOME desktop in a container. It's not that hard. I was mislead to take a harder solution.
